parsers/openapi.c

```diff
@@ -0,0 +1,363 @@
+	{ true, 'd', "schema", "schemas" },
+	{ true, 'p', "path", "paths" },
+	{ true, 'R', "response", "responses" },
+	{ true, 'P', "parameter", "parameters" },
+};
+
+enum openapiKeys {
+	KEY_UNKNOWN,
+	KEY_PATHS,
+	KEY_COMPONENTS,
+	KEY_SCHEMAS,
+	KEY_PARAMETERS,
+	KEY_RESPONSES,
+	KEY_DEFINITIONS,
+};
+
+struct yamlBlockTypeStack {
+	yaml_token_type_t type;
+	enum openapiKeys key;
+	struct yamlBlockTypeStack *next;
+};
+
+/* - name: "THE NAME" */
+enum openapiPlayDetectingState {
+	DSTAT_LAST_KEY,
+	DSTAT_LAST_VALUE,
+	DSTAT_INITIAL,
+};
+
+
+struct sOpenApiSubparser {
+	yamlSubparser yaml;
+	struct yamlBlockTypeStack *type_stack;
+	enum openapiPlayDetectingState play_detection_state;
+};
+
+static void pushBlockType (struct sOpenApiSubparser *openapi, yaml_token_type_t t)
+{
+	struct yamlBlockTypeStack *s;
+
+	s = xMalloc (1, struct yamlBlockTypeStack);
+
+	s->next = openapi->type_stack;
+	openapi->type_stack = s;
+
+	s->type = t;
+	s->key = KEY_UNKNOWN;
+}
+
+static void popBlockType (struct sOpenApiSubparser *openapi,
+						  yaml_token_t *token)
+{
+	struct yamlBlockTypeStack *s;
+
+	s = openapi->type_stack;
+	openapi->type_stack = s->next;
+
+	s->next = NULL;
+
+	eFree (s);
+}
+
+static void popAllBlockType (struct sOpenApiSubparser *openapi,
+							 yaml_token_t *token)
+{
+	while (openapi->type_stack)
+		popBlockType (openapi, token);
+}
+
+static bool stateStackMatch (struct yamlBlockTypeStack *stack,
+							 const enum openapiKeys *expectation,
+							 unsigned int length
+							 )
+{
+	if (length == 0)
+	{
+		if (stack == NULL)
+			return true;
+		else
+			return false;
+	}
+
+	if (stack == NULL)
+		return false;
+
+	if (stack->key == expectation[0])
+		return stateStackMatch (stack->next, expectation + 1, length - 1);
+	else
+		return false;
+}
+
+static bool scalarNeq (yaml_token_t *token, unsigned int len, const char* val)
+{
+	if ((token->data.scalar.length == len)
+		&& (strncmp (val, (char *)token->data.scalar.value, len) == 0))
+		return true;
+	else
+		return false;
+}
+
+static enum openapiKeys parseKey(yaml_token_t *token)
+{
+  if (scalarNeq(token, 10, "components"))
+	  return KEY_COMPONENTS;
+  else if (scalarNeq(token, 7, "schemas"))
+	  return KEY_SCHEMAS;
+  else if (scalarNeq(token, 5, "paths"))
+	  return KEY_PATHS;
+  else if (scalarNeq(token, 11, "definitions"))
+	  return KEY_DEFINITIONS;
+  else if (scalarNeq(token, 10, "parameters"))
+	  return KEY_PARAMETERS;
+  else if (scalarNeq(token, 9, "responses"))
+	  return KEY_RESPONSES;
+  else
+	  return KEY_UNKNOWN;
+}
+
+static void printStack(struct yamlBlockTypeStack* stack)
+{
+  if (!stack) {
+	  puts("");
+	  return;
+  }
+
+  printf("[%d] - ", stack->key);
+  printStack(stack->next);
+}
+
+
+struct tagSource {
+	openapiKind kind;
+	const enum openapiKeys* keys;
+	size_t countKeys;
+};
+
+static const enum openapiKeys pathKeys[] = {
+	KEY_UNKNOWN,
+	KEY_PATHS,
+};
+
+static const enum openapiKeys responses3Keys[] = {
```

Comment:
> Is there any difference between the targets, Swagger 2.0 and Openapi 3.0.x?

They are 2 different revisions of the protocol (Swagger == OpenAPI 2.0). They are mostly the same except for several details (not interesting for ctags) and several sections have a distinct path in yaml (e.g. "responses" in swagger and "components.responses" for openapi). I tried to express it in `XXXKeys` names with 2 (swagger) or 3 (openapi). `pathKeys` is the same for both.

Reply:
I see. Thank you.